anesthetic/plot.py

```diff
@@ -566,22 +605,31 @@ def kde_contour_plot_2d(ax, data_x, data_y, *args, **kwargs):
 
     contours = iso_probability_contours_from_samples(p, weights=w)
 
-    cmap = kwargs.pop('cmap', basic_cmap(color))
+    if facecolor not in [None, 'None', 'none']:
+        linewidths = kwargs.pop('linewidths', 0.5)
+        cmap = kwargs.pop('cmap', basic_cmap(facecolor))
+        contf = ax.tricontourf(tri, p, contours, cmap=cmap, zorder=zorder,
+                               vmin=0, vmax=p.max(), *args, **kwargs)
+        for c in contf.collections:
+            c.set_cmap(cmap)
+        ax.patches += [plt.Rectangle((0, 0), 0, 0, fc=cmap(0.999),
+                                     ec=cmap(0.32), lw=2, label=label)]
+        cmap = None
+    else:
+        contf = None
+        cmap = kwargs.pop('cmap', None)
+        edgecolor = edgecolor if cmap is None else None
+        linewidths = kwargs.pop('linewidths', 1.5)
+        ax.patches += [plt.Rectangle((0, 0), 0, 0, fc=None, ec=edgecolor,
+                                     lw=2, label=label)]
 
-    cbar = ax.tricontourf(tri, p, contours, cmap=cmap,
-                          zorder=zorder, vmin=0, vmax=p.max(), *args, **kwargs)
-    for c in cbar.collections:
-        c.set_cmap(cmap)
-
-    ax.tricontour(tri, p, contours, zorder=zorder,
-                  vmin=0, vmax=p.max(), linewidths=linewidths, colors='k',
-                  *args, **kwargs)
-    ax.patches += [plt.Rectangle((0, 0), 0, 0, fc=cmap(0.999), ec=cmap(0.32),
-                                 lw=2, label=label)]
+    cont = ax.tricontour(tri, p, contours, zorder=zorder,
+                         vmin=0, vmax=0.32 * p.max(), linewidths=linewidths,
+                         colors=edgecolor, cmap=cmap, *args, **kwargs)
```

Comment:
I was surprised at very faint contour lines when using `cmap`. Turns out this had to do with `vmax=p.max()`. In the filled case, this is not noticeable, because `vmax` is ignored when `colors` is a constant. Setting `vmax=0.32 * p.max()` sets the color of the inner contour line to the extreme end of `cmap`. Happy with this behaviour?

Reply:
The hard coded `0.32` are due to the hard coded `contours=[0.68, 0.95]` in `iso_probability_contours_from_samples`.

Reply:
>I was surprised at very faint contour lines when using cmap

Could you provide a picture to show this behaviour, I'm not sure exactly what you mean

Reply:
Both following plots were generated with:

```python
from matplotlib.colors import LinearSegmentedColormap as LSC
ns = NestedSamples(root=anesthetic/tests/example_data/pc")
fig = plt.figure(figsize=(8, 8))
fig, axes = make_2d_axes(['x0', 'x1', 'x2', 'x3', 'x4'], fig=fig)
ns.plot_2d(axes, fc=None, c='C0', cmap=LSC.from_list('', ['w', 'C0']));
```

This is with `vmax=0.32*p.max()` (new behaviour):

![unfilled](https://user-images.githubusercontent.com/18258042/88593963-4580f480-d058-11ea-95d0-eac4d0d50012.png)

This is with `vmax=p.max()` (old behaviour):

![unfilled](https://user-images.githubusercontent.com/18258042/88594335-eb346380-d058-11ea-9e73-bb6471eec536.png)

Reply:
Note, that this behaviour also applies to the filled contours. If you look closely, the inner contour does not have the same saturated hue as lines in the 1d plots. For the filled contours, I think this is actually desirable, but for line contours this seems a bit too washed out.

What is special in our case is that we have only 2 levels, which is not the typical usecase of these contour plots.

Reply:
This makes it a lot clearer.

Reply:
Hmm, there seems to be a more general problem with the calculation of `contours` in `kde_plot_2d`. Can you see how  the outer contour lines in `x0-x3`, `x1-x3` and `x3-x4` are darker than in the other plots? This is independent of filled/unfilled contours (see e.g. plot below comparing black to blue edges), but also happens in the current master, just less noticeable because the black edges hide the (lacking) contrast to the white background.

I've looked at what contours computes and got:
```
01, contours=[2.000, 10.996, 34.730],    level2 / level3 = 0.317,    level1 / level3 = 0.058
02, contours=[0.959,  7.355, 28.050],    level2 / level3 = 0.262,    level1 / level3 = 0.034
12, contours=[0.788,  6.859, 26.510],    level2 / level3 = 0.259,    level1 / level3 = 0.030
03, contours=[0.693,  2.406,  4.204],    level2 / level3 = 0.572,    level1 / level3 = 0.165   *
13, contours=[0.730,  2.415,  4.095],    level2 / level3 = 0.590,    level1 / level3 = 0.178   *
23, contours=[0.477,  3.865,  8.169],    level2 / level3 = 0.473,    level1 / level3 = 0.058
04, contours=[0.319,  1.178,  4.240],    level2 / level3 = 0.278,    level1 / level3 = 0.075
14, contours=[0.288,  1.176,  4.072],    level2 / level3 = 0.289,    level1 / level3 = 0.071
24, contours=[0.240,  1.974,  7.138],    level2 / level3 = 0.277,    level1 / level3 = 0.034
34, contours=[0.244,  0.563,  1.234],    level2 / level3 = 0.456,    level1 / level3 = 0.198   *
```
Note the lines marked with `*` deviate from the expected `0.32` and `0.05` values associated with 68% and 95% contours.

Is maybe something wrong in `anesthetic.utils.iso_probability_contours_from_samples`? Or are outliers such as these expected?

Reply:
Maybe related, I've noticed that the maximum width of the posterior contours in the `*` marked axes are a bit slimmer than in the others. Wouldn't you expect the maximum width to be roughly consistent across axes?

Reply:
This difference in colouration is a feature rather than a bug -- it's actually quietly mentioned on the front page of the repo:

>Many KDE plotting tools have conventions that don't play well with uniformly distributed parameters, which presents a problem if you are trying to plot priors along with your posteriors. anesthetic has a sensible mechanism, by defining the contours by the amount of iso-probability mass they contain, but colouring the fill in relation to the probability density of the contour.

This makes a big difference when plotting priors. For an example, have a look at the blue or orange contours in [Figure 4](https://arxiv.org/pdf/1908.09139v1.pdf). If these were coloured in accordance with mass, rather than density, the picture would misleadingly look like a constraint.

Reply:
However, I agree that the contour lines themselves, if not coloured `'k'`, should be the same colour as the 68% region.

Reply:
> However, I agree that the contour lines themselves, if not coloured `'k'`, should be the same colour as the 68% region.

Except they weren't. My thingy was an ugly hack. I have now improved things, such that the colours from `contour` match the colours from `contourf` when using a colour map. The problem was that `contourf` assigns colours to the region inbetween levels, whereas `contour` assigns colours to the level itself...

Sorry for yet another review request... I think I've got it all now.